src/mongocrypt-ctx-encrypt.c

```diff
@@ -244,42 +203,84 @@ _fle2_mongo_op_markings (mongocrypt_ctx_t *ctx, mongocrypt_binary_t *out)
    }
+   bson_copy_to (&cmd_bson, out);
 
    if (!_mongocrypt_buffer_to_bson (&ectx->encrypted_field_config,
                                     &encrypted_field_config_bson)) {
       return _mongocrypt_ctx_fail_w_msg (
          ctx, "unable to convert encrypted_field_config to BSON");
    }
 
-   bson_copy_to (&cmd_bson, &mongocryptd_cmd_bson);
-   if (!_fle2_append_encryptionInformation (&mongocryptd_cmd_bson,
-                                            ectx->ns,
-                                            &encrypted_field_config_bson,
-                                            ctx->status)) {
+   if (!_fle2_append_encryptionInformation (
+          out, ectx->ns, &encrypted_field_config_bson, ctx->status)) {
       return _mongocrypt_ctx_fail (ctx);
    }
 
-   _mongocrypt_buffer_steal_from_bson (&ectx->mongocryptd_cmd,
-                                       &mongocryptd_cmd_bson);
+   return true;
+}
 
-   out->data = ectx->mongocryptd_cmd.data;
-   out->len = ectx->mongocryptd_cmd.len;
+
+/**
+ * @brief Create the server-side command that contains information for
+ * generating encryption markings via query analysis.
+ *
+ * @param ctx The encryption context.
+ * @param out The destination of the generate BSON document
+ * @return true On success
+ * @return false Otherwise. Sets a failing status message in this case.
+ */
+static bool
+_create_markings_cmd_bson (mongocrypt_ctx_t *ctx, bson_t *out)
+{
+   _mongocrypt_ctx_encrypt_t *ectx = (_mongocrypt_ctx_encrypt_t *) ctx;
+   if (!_mongocrypt_buffer_empty (&ectx->encrypted_field_config)) {
+      // Defer to FLE2 to generate the markings command
+      return _fle2_mongo_op_markings (ctx, out);
+   }
+
+   // For FLE1:
+   // Get the original command document
+   bson_t bson_view = BSON_INITIALIZER;
+   if (!_mongocrypt_buffer_to_bson (&ectx->original_cmd, &bson_view)) {
+      _mongocrypt_ctx_fail_w_msg (ctx, "invalid BSON cmd");
+      return false;
+   }
+   // Copy the command to the output
+   bson_copy_to (&bson_view, out);
+
+   if (!_mongocrypt_buffer_empty (&ectx->schema)) {
+      // We have a schema buffer. View it as BSON:
+      if (!_mongocrypt_buffer_to_bson (&ectx->schema, &bson_view)) {
+         _mongocrypt_ctx_fail_w_msg (ctx, "invalid BSON schema");
+         return false;
+      }
+      // Append the jsonSchema to the output command
+      BSON_APPEND_DOCUMENT (out, "jsonSchema", &bson_view);
+   }
+
+   // if a local schema was not set, set isRemoteSchema=true
+   BSON_APPEND_BOOL (out, "isRemoteSchema", !ectx->used_local_schema);
    return true;
 }
 
+
 static bool
 _mongo_op_markings (mongocrypt_ctx_t *ctx, mongocrypt_binary_t *out)
 {
-   _mongocrypt_ctx_encrypt_t *ectx;
+   _mongocrypt_ctx_encrypt_t *ectx = (_mongocrypt_ctx_encrypt_t *) ctx;
 
-   ectx = (_mongocrypt_ctx_encrypt_t *) ctx;
    if (_mongocrypt_buffer_empty (&ectx->mongocryptd_cmd)) {
+      // We need to generate the command document
       bson_t cmd_bson;
```

Comment:
```suggestion
      bson_t cmd_bson = BSON_INITIALIZER;
```

If L245 returns,out may be left uninitialized.

If L255 returns false, out may be initialized to a non-empty BSON document. The caller needs to call bson_destroy to prevent leaking. bson_t heap allocates when the size of the BSON document exceeds 120 bytes.

Initializing `cmd_bson` is consistent with the other call to `_create_markings_cmd_bson`.